libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,907 @@
+    ///
+    /// [`MAX_PAGE_SIZE`]: constant.MAX_PAGE_SIZE.html
+    /// [`MAX_PAGE`]: constant.MAX_PAGE.html
+    /// [`MAX_CYCLE`]: constant.MAX_CYCLE.html
+    fn is_storage_supported<S: Storage>(storage: &S) -> bool {
+        let word_size = storage.word_size();
+        let page_size = storage.page_size();
+        let num_pages = storage.num_pages();
+        let max_word_writes = storage.max_word_writes();
+        let max_page_erases = storage.max_page_erases();
+        word_size == 4
+            && page_size % word_size == 0
+            && (8 * word_size <= page_size && page_size <= MAX_PAGE_SIZE)
+            && (3 <= num_pages && num_pages <= MAX_PAGE + 1)
+            && max_word_writes >= 2
+            && max_page_erases <= MAX_CYCLE
+    }
+
+    /// The size of a word in bytes.
+    pub fn word_size(&self) -> usize {
+        WORD_SIZE
+    }
+
+    /// The size of a page in bytes.
+    ///
+    /// We have `32 <= self.page_size() <= MAX_PAGE_SIZE`.
+    pub fn page_size(&self) -> usize {
+        self.page_size
+    }
+
+    /// The number of pages in the storage.
+    ///
+    /// Notation: `N`. We have `3 <= N <= MAX_PAGE + 1`.
+    pub fn num_pages(&self) -> usize {
+        self.num_pages
+    }
+
+    /// The maximum page index.
+    ///
+    /// We have `2 <= self.max_page() <= MAX_PAGE`.
+    pub fn max_page(&self) -> usize {
+        self.num_pages - 1
+    }
+
+    /// The maximum number of times a page can be erased.
+    ///
+    /// Notation: `E`. We have `E <= MAX_CYCLE`.
+    pub fn max_page_erases(&self) -> usize {
+        self.max_page_erases
+    }
+
+    /// The maximum key.
+    pub fn max_key(&self) -> usize {
+        MAX_KEY
+    }
+
+    /// The maximum number of updates per transaction.
+    pub fn max_updates(&self) -> usize {
+        MAX_UPDATES
+    }
+
+    /// The size of a virtual page in words.
+    ///
+    /// A virtual page is stored in a physical page after the page header.
+    ///
+    /// Notation: `Q`. We have `6 <= Q <= MAX_VIRT_PAGE_SIZE`.
+    pub fn virt_page_size(&self) -> usize {
+        self.page_size() / self.word_size() - CONTENT_WORD
+    }
+
+    /// The maximum length in bytes of a user payload.
+    ///
+    /// We have `20 <= self.max_value_len() <= MAX_VALUE_LEN`.
+    pub fn max_value_len(&self) -> usize {
+        min(
+            (self.virt_page_size() - 1) * self.word_size(),
+            MAX_VALUE_LEN,
+        )
+    }
+
+    /// The maximum prefix length in words.
+    ///
+    /// A prefix is the first words of a virtual page that belong to the last entry of the previous
+    /// virtual page. This happens because entries may overlap up to 2 virtual pages.
+    ///
+    /// Notation: `M`. We have `5 <= M < Q`.
+    pub fn max_prefix_len(&self) -> usize {
+        self.bytes_to_words(self.max_value_len())
+    }
+
+    /// The total virtual capacity in words.
+    ///
+    /// Notation: `V`. We have `V = (N - 1) * (Q - 1) - M`.
```

Comment:
I don't see any check directly involving `Q` in `is_storage_supported`. Not saying this isn't implied by the checks in `is_storage_supported`, but it's not a trivial conclusion either.

Reply:
I think there 2 possible ways to understand your initial question and I'm not sure which one it is:
1. Being worried that the subtraction may underflow. Rust checks in debug mode so unit tests (which are in upcoming `store.rs`) and fuzzing are going to catch it.
2. Being concerned about proving that it will not underflow. The proof is simple enough (using existing inequalities provided in other functions documentation) to not be useful to write (compared to proving that the store is correct which is not even tractable on paper which is why fuzzing is necessary, but I still sketched the main ideas in `lib.rs`).

Reply:
Resolving here to avoid multiplying threads, but the same remark applies: if the proof is easy, can you write down & document it?